Fixes Session timeout error #2896 and #2897

sancsin opened the pull request:
<!--
Thank you for contributing to Lychee! We only accept PR to the master branch.

In addition, please describe the benefit to end users; the reasons it does not break any existing features, etc.

If you're pushing a Feature:
- Title it: "This new feature"
- Describe what the new feature enables
- Add tests to test the new feature.
- Ensure it doesn't break any existing features.

If you're pushing a Fix:
- Title it: "Fixes the bug name"
- If it is a fix an an existing issue start the description with `fixes #xxxx` where xxxx is the issue number.
- Describe how it fixes in a few words.
- Add a test that triggered the bug before fixing it.
- Ensure it doesn't break any feature.

All Pull Requests run with extensive tests for stable and latest versions of PHP. 
Ensure your tests pass or your PR may be taken down.

Additionally you can run `make phpstan` and `make formatting` on your own machine as they will be required to pass for your PR to be accepted.

Don't worry if your code styling isn't perfect! php-cs-fixer will automatically create a pull request with any style fixes. This allows us to focus on the content of the contribution and not the code style.
-->

Fixes #2896 and #2897 by providing clear feedback to the user. The fix does so by using a middleware that creates a cookie to compare the elapsed time since a user last logged-in and the session lifetime configuration setting. If elapsed time exceeds the session lifetime, a SessionExpiredException is thrown. This error is also caught in the front-end through axios interceptor to inform the user using an alert that the session has expired. Once the user clicks on "OK", the page reloads. Reloading page provides expected behaviour. For example, if the user clicks on an album that is public, the page will reload and the now guest user, will still be able to browse the public album. However, if the user clicks on an album that is not public, the page will reload and provide clear exception feedback to the user that they are not authorized to view the album/page since their session expired.

Another scenario that this fix provides better user experience for is when the user is on a navigated page that is provided through the  left menu. Once the session expires and the now guest user tries to access any of the pages that they should not have access to, they are clearly provided with the session expired alert. Furthermore, the page reloads instead of showing errors that don't make sense for the user, such as the ones described in #2897.

sancsin committed:
Adding CheckSessionTimeout Middleware

sancsin committed:
Better error handling

sancsin committed:
Better documentation and adding new HTTP status code

sancsin committed:
Merge branch 'master' of https://github.com/sancsin/Lychee into redirect_home_on_session_expiration

ildyria committed:
Dropping support of php8.2 (#2884)

ildyria committed:
Update artifact actions (#2885)

sancsin committed:
Handling some edge cases for session expiration

ildyria committed:
Improve version wording (#2887)

* improve version wording

* Update resources/views/install/requirements.blade.php

Co-authored-by: Martin Stone <1611702+d7415@users.noreply.github.com>

---------

Co-authored-by: Martin Stone <1611702+d7415@users.noreply.github.com>

sancsin committed:
Reload instead of redirect approach

Reload approach seems more user friendly as it handles loading of
publicly available albums better

Also removed unnecessary log calls

ildyria committed:
Reset localization (sorry) + improved clarity of language setup (#2877)



---------

Co-authored-by: Martin Stone <1611702+d7415@users.noreply.github.com>

ildyria committed:
UI fix maintenance button alignment (#2888)

ildyria committed:
Use h-40 instead of h-56 for maintenance block. (#2893)

ildyria committed:
Fix diagnostics not being complete on version 6 (#2892)

ildyria committed:
Add Docker info, do not display Update Maintenance on docker (#2894)

sancsin committed:
CheckSessionTimeout and reload page refinement

sancsin committed:
Using translation function for the error message

sancsin committed:
Removed unnecessary exception mapping

ildyria commented on lang/en/lychee.php:
```diff
@@ -523,6 +523,7 @@
 	'ERROR_ALBUM_NOT_FOUND' => 'Error: album %s not found',
 	'ERROR_DROPBOX_KEY' => 'Error: Dropbox key not set',
 	'ERROR_SESSION' => 'Session expired.',
+	'ERROR_SESSION_RELOAD' => 'Session expired. Click OK to reload the page',
```
This file (`lychee.php`) is deprecated.
Because this is related to error messages, I think it might be best to create a new `error.php` translation file where this line is contained or add it to `settings.php` maybe?

Note that we have a translation check, so that new file/line will also need to be added to all the other languages.
It does not need to be translated, but just as a place holder with the untranslated string.

sancsin replied:
Where is `settings.php`? Can't find it anywhere in the codebase. Does it need to created?

ildyria replied:
https://github.com/LycheeOrg/Lychee/blob/master/lang/en/settings.php

sancsin replied:
Aha..Looks like I hadn't synced the fork in a while, that is why I couldn't see the `settings.php` file. Thanks!

sancsin replied:
Have added the string to a file called `common.php`. Can maybe be used for translated strings that are used commonly across Lychee?

sancsin replied:
There seems to be 1 test that is failing. I'm not sure what the problem is. @ildyria could you check and assist here when you have the possibility?

ildyria replied:
> Cannot parse /home/runner/work/Lychee/Lychee/app/Http/Controllers/VueController.php: Syntax error, unexpected T_SL on line 56

Yeah that one, I have no clue either... It always happens on external pull requests...

sancsin committed:
Correcting typo and fixing phpstan errors

sancsin committed:
Fixing php code style error

sancsin committed:
Merge branch 'LycheeOrg:master' into session_expiration_feedback

sancsin committed:
Handled translation string

Handled translation string according to comment from reviewer

Commit:
Fixing check errors

Commit:
Fixing PHP code style errors

ildyria commented on app/Http/Controllers/VueController.php:
```diff
@@ -53,8 +53,12 @@ public function view(?string $albumId = null, ?string $photoId = null): View
 				Gate::authorize(PhotoPolicy::CAN_SEE, [Photo::class, $photo]);
 				session()->now('photo', $photo);
 			}
-		} catch (ModelNotFoundException) {
```
Is this change necessary ?

ildyria replied:
I don't think so.

sancsin replied:
Yes, you are right. This is not necessary. Since it is only `$albumFactory->findAbstractAlbumOrFail($albumId, false);` that is throwing an error.

sancsin replied:
Have made a new commit and restored these lines.

ildyria commented on resources/js/config/axios-config.ts:
```diff
@@ -11,8 +22,12 @@ const AxiosConfig = {
 			function (config: AxiosRequestConfig) {
-				const token = CSRF.get();
-				(config.headers as AxiosRequestHeaders)["X-XSRF-TOKEN"] = token;
+				try {
```
what error are you catching here ?

ildyria replied:
okay I figured it out...

sancsin replied:
It is the following error that I'm catching. Which happens after session has timed out. If we don't catch it then we get an uncaught exception in the browser console.

![Screenshot_20250112_223716](https://github.com/user-attachments/assets/da67467a-b9d0-4282-9854-e2485837c8cd)

ildyria replied:
Yep I saw that. I made quite a few changes to your implementation, can you check if that is working satisfactory for you ? :)

sancsin replied:
Looks quite more like an elegant solution! :)

sancsin committed:
Removed unnecessary catch exception

ildyria committed:
No middleware necessary + modal instead of alert

The try catch block in axios got me thinking. The cookie is set to
expire with the session. As a result we can track this to know whether
it has expired or not.

Trigger a specific event which is capture by the Error page and display
a modal instead of a default JS alert. This is UI and UX friendlier.

ildyria committed:
less logs

ildyria reviewed (approved):
LGTM, waiting for double check from @sancsin

sancsin commented:
The _Session Expired_ dialog appears even though user is not logged-in and tries to open a publicly available album. Is that the desired behaviour?

sancsin commented:
At the moment, the user seems to be redirected to `http://address/gallery/gallery` instead of `http://address/gallery/` when clicking on _Go to the Gallery_. Furthemore, the dialog does not close when clicking on _Go to the Gallery_ button.

ildyria commented:
> The _Session Expired_ dialog appears even though user is not logged-in and tries to open a publicly available album. Is that the desired behaviour?

definitely not.

sancsin commented:
Seeing the below errors when navigating to "Settings"
![Screenshot_20250112_234250](https://github.com/user-attachments/assets/2b426ce3-fb30-4f7f-8181-a431eb6e7b46)

ildyria commented:
That is so weird, because I do not have that same behaviour here. O.o

ildyria commented:
Wait... Are you on http or https ? 
Though it shouldn't matter. This is tied to the xsrf token.

ildyria committed:
fix routes

sancsin commented:
> Wait... Are you on http or https ? Though it shouldn't matter. This is tied to the xsrf token.

I'm on http

ildyria commented:
So I see what you mean with the dialog popping up if you e.g. open an album while you spent too much time on the page...

ildyria commented:
So I figured it out.

The CSRF check is as follows:
```php
    public function handle($request, Closure $next)
    {
        if (
            $this->isReading($request) ||
            $this->runningUnitTests() ||
            $this->inExceptArray($request) ||
            $this->tokensMatch($request)
        ) {
            return tap($next($request), function ($response) use ($request) {
                if ($this->shouldAddXsrfTokenCookie()) {
                    $this->addCookieToResponse($request, $response);
                }
            });
        }

        throw new TokenMismatchException('CSRF token mismatch.');
    }
```
The `TokenMismatchException` was responsible for the 419 error.

The function `isReading()` is the important one here. As opposed to version 4 and 5 where all requests were done in `POST`, the v6 makes use also of other verbs such as `DELETE`, `PATCH`, `PUT` and more importantly `GET`.

Following the code path, we see the two following important functions:
```php
    protected function isReading($request)
    {
        return in_array($request->method(), ['HEAD', 'GET', 'OPTIONS']);
    }
```
and this (which previously was always used).
```php
    protected function tokensMatch($request)
    {
        $token = $this->getTokenFromRequest($request);

        return is_string($request->session()->token()) &&
               is_string($token) &&
               hash_equals($request->session()->token(), $token);
    }
```
**Conclusion**: GET, HEAD, and OPTION request do not require the XSRF token.

But if your session times out, and try to access a protected folder you will get a 401.
We took care of the exception which was triggering the _white_ screen.

Note that your original middleware idea was leading to the same result as the behaviour you observed: triggering a error and refresh request but with the use of an extra cookie.

Now the question: **Should we have the XSRF token made mandatory for the GET requests of v6 or not?**
If yes, then the current implementation checking the token absence or not is the way to go with the client side trigger of the modal.

If not, then this requires a more difficult handling of 401 errors.

ildyria committed:
force session on GET api/v2/

ildyria commented:
@sancsin see latest commit, in the behavior change. :)

sancsin commented:
> So I figured it out.
> 
> The CSRF check is as follows:
> 
> ```
>     public function handle($request, Closure $next)
>     {
>         if (
>             $this->isReading($request) ||
>             $this->runningUnitTests() ||
>             $this->inExceptArray($request) ||
>             $this->tokensMatch($request)
>         ) {
>             return tap($next($request), function ($response) use ($request) {
>                 if ($this->shouldAddXsrfTokenCookie()) {
>                     $this->addCookieToResponse($request, $response);
>                 }
>             });
>         }
> 
>         throw new TokenMismatchException('CSRF token mismatch.');
>     }
> ```
> 
> The `TokenMismatchException` was responsible for the 419 error.
> 
> The function `isReading()` is the important one here. As opposed to version 4 and 5 where all requests were done in `POST`, the v6 makes use also of other verbs such as `DELETE`, `PATCH`, `PUT` and more importantly `GET`.
> 
> Following the code path, we see the two following important functions:
> 
> ```
>     protected function isReading($request)
>     {
>         return in_array($request->method(), ['HEAD', 'GET', 'OPTIONS']);
>     }
> ```
> 
> and this (which previously was always used).
> 
> ```
>     protected function tokensMatch($request)
>     {
>         $token = $this->getTokenFromRequest($request);
> 
>         return is_string($request->session()->token()) &&
>                is_string($token) &&
>                hash_equals($request->session()->token(), $token);
>     }
> ```
> 
> **Conclusion**: GET, HEAD, and OPTION request do not require the XSRF token.
> 
> But if your session times out, and try to access a protected folder you will get a 401. We took care of the exception which was triggering the _white_ screen.
> 
> Note that your original middleware idea was leading to the same result as the behaviour you observed: triggering a error and refresh request but with the use of an extra cookie.
> 
> Now the question: **Should we have the XSRF token made mandatory for the GET requests of v6 or not?** If yes, then the current implementation checking the token absence or not is the way to go with the client side trigger of the modal.
> 
> If not, then this requires a more difficult handling of 401 errors.

Thanks for the detailed explanation. Really appreciate it. Now I understand a little better. I think the behaviour now is quite good even though the session timeout dialog is shown when the user is logged-out and tries to access public album. I don't know if you noticed though, with the cookie approach this issue was not there.

sancsin committed:
Close SessionExpiredReload dialog on routing to gallery

sancsin commented:
> @sancsin see latest commit, in the behavior change. :)

I made a new commit as a suggestion to handle the cases where the SessionExpiredReload dialog was not closing when user clicks on _Go to gallery_ button.

sancsin committed:
Formatting fix

ildyria committed:
simplify
